src/Lean/Elab/Syntax.lean

```diff
@@ -581,69 +583,83 @@ builtin_initialize
-  let stxParts := #[stxPart] ++ stxParts
-  -- name
-  let name ← match name? with
-    | some name => pure name
-    | none => liftMacroM <| mkNameFromParserSyntax cat.getId (mkNullNode stxParts)
-  -- build pattern for syntax `match`
-  let patHead ← liftMacroM <| expandMacroArgIntoPattern head
-  let patArgs ← liftMacroM <| args.mapM expandMacroArgIntoPattern
-  let pat := Syntax.node (currNamespace ++ name) (#[patHead] ++ patArgs)
-  let stxCmd ← `(Parser.Command.syntax|
-    $attrKind:attrKind syntax $(prec)? (name := $(mkIdentFrom stx name):ident) (priority := $(quote prio):numLit) $[$stxParts]* : $cat)
-  let elabCmd ←
-    if expectedTypeSpec.hasArgs then
-      if catName == `term then
-        let expId := expectedTypeSpec[1]
-        `(@[termElab $(mkIdentFrom stx name):ident] def elabFn : Lean.Elab.Term.TermElab :=
-          fun stx expectedType? => match stx with
-            | `($pat) => Lean.Elab.Command.withExpectedType expectedType? fun $expId => $rhs
-            | _ => throwUnsupportedSyntax)
+def elabElabRulesAux (doc? : Option Syntax) (attrKind : Syntax) (k : SyntaxNodeKind) (cat? expty? : Option Syntax) (alts : Array Syntax) : CommandElabM Syntax := do
+  let alts ← alts.mapM fun alt => match alt with
+    | `(matchAltExpr| | $pats,* => $rhs) => do
+      let pat := pats.elemsAndSeps[0]
+      if !pat.isQuot then
+        throwUnsupportedSyntax
+      let quoted := getQuotContent pat
+      let k' := quoted.getKind
+      if checkRuleKind k' k then
+        pure alt
+      else if k' == choiceKind then
+         match quoted.getArgs.find? fun quotAlt => checkRuleKind quotAlt.getKind k with
+         | none        => throwErrorAt alt "invalid elab_rules alternative, expected syntax node kind '{k}'"
+         | some quoted =>
+           let pat := pat.setArg 1 quoted
+           let pats := pats.elemsAndSeps.set! 0 pat
+           `(matchAltExpr| | $pats,* => $rhs)
       else
-        throwErrorAt expectedTypeSpec "syntax category '{catName}' does not support expected type specification"
-    else if catName == `term then
-      `(@[termElab $(mkIdentFrom stx name):ident] def elabFn : Lean.Elab.Term.TermElab :=
-        fun stx _ => match stx with
-          | `($pat) => $rhs
-          | _ => throwUnsupportedSyntax)
-    else if catName == `command then
-      `(@[commandElab $(mkIdentFrom stx name):ident] def elabFn : Lean.Elab.Command.CommandElab :=
-        fun
-          | `($pat) => $rhs
-          | _ => throwUnsupportedSyntax)
-    else if catName == `tactic then
-      `(@[tactic $(mkIdentFrom stx name):ident] def elabFn : Lean.Elab.Tactic.Tactic :=
-        fun
-          | `(tactic|$pat) => $rhs
-          | _ => throwUnsupportedSyntax)
+        throwErrorAt alt "invalid elab_rules alternative, unexpected syntax node kind '{k'}'"
+    | _ => throwUnsupportedSyntax
+  let catName ← match cat?, expty? with
+    | some cat, _ => cat.getId
+    | _, some _   => `term
+    -- TODO
```

Comment:
Please add a comment explaining what is missing.

Reply:
Done. I'm not quite sure what is reasonable here, do we want a mapping from syntax kinds to categories?

Reply:
Btw, I notice I started calling syntax node kinds syntax kinds since the former is a mouthful and we extended `Syntax.getKind` to make sense on any `Syntax` constructor. What do you think?